### PR TITLE
Implement SyntheticBanditDataWithActionEmbeds and Marginalized IPS

### PR DESCRIPTION
**Pull request by usaito**
## Overview
- Implement `obp.ope.MarginalziedInverseProbabilityWeighting` and `obp.ope.SelfNormalizedMarginalizedInverseProbabilityWeighting`. The first one is the main proposal of the reference paper (called MIPS in that paper), and the latter is a trivial extension.
- Implement `obp.dataset.SyntheticBanditDatasetWithActionEmbeds`, which generates synthetic bandit data with action embeddings consisting of some discrete random variables (e.g., category information of movies). The data generating process of this class follows the synthetic experiment conducted in the reference paper.

## Reference
Yuta Saito and Thorsten Joachims.
"Off-Policy Evaluation for Large Action Spaces via Embeddings.", 2022.
https://arxiv.org/abs/2202.06317

### CODE REVIEW

File: obp/ope/estimators_embed.py [hunk 1]
```diff
@@ -0,0 +1,666 @@
...
+            for idx in idx_list:
+                exclude_d = d_list_[idx]
+                theta_i, cnf_i = theta_list_[idx], cnf_list_[idx]
+                theta_j, cnf_j = np.array(theta_list), np.array(cnf_list)
+                if (np.abs(theta_j - theta_i) <= cnf_i + C * cnf_j).all():
+                    theta_list.append(theta_i), cnf_list.append(cnf_i)
+                else:
+                    exclude_d_idx = np.where(current_feat != exclude_d, True, False)
+                    return theta_j[-1]
+            exclude_d_idx = np.where(current_feat != exclude_d, True, False)
```
Comment (fullflu) [inline obp/ope/estimators_embed.py:421]: [ask]

This line seems to exclude a feature that has the smallest CDF.
Is this a reasonable algorithm?
(Intuitively, the smallest lower bound feature has some reason to be excluded.)

File: obp/ope/estimators_embed.py [hunk 2]
```diff
@@ -0,0 +1,666 @@
...
+                if len(theta_list) > 0:
+                    theta_list_.append(theta), cnf_list_.append(cnf)
+                else:
+                    theta_list.append(theta), cnf_list.append(cnf)
+                    continue
+
+            idx_list = np.argsort(cnf_list_)[::-1]
+            for idx in idx_list:
+                theta_i, cnf_i = theta_list_[idx], cnf_list_[idx]
+                theta_j, cnf_j = np.array(theta_list), np.array(cnf_list)
```
Comment (fullflu) [inline obp/ope/estimators_embed.py:356]: [ask]

When `len(theta_list) == 1`, this function could return `theta_list[0]` even though `Eq. (22)` does not hold.

Is this an intended behavior?

I wonder if the initial value of `theta_list` should be the maximum value of `cnt_list_` of index `i`.

**Review by fullflu — COMMENTED** [review]
I'm sorry for the late review...!

My 1st review mainly focuses on the estimators.

In the 2nd review, I will check the testing details.

File: obp/ope/estimators_embed.py (hunk 1 above)
Comment (usaito) [reply obp/ope/estimators_embed.py:421]: Yes, this is intended. `theta_j[-1]` is the last element of `theta_j` (or equivalently `theta_list`), this is not necessarily the element having the smallest CDF.

File: obp/ope/estimators_embed.py (hunk 2 above)
Comment (usaito) [reply obp/ope/estimators_embed.py:356]: `len(theta_list) == 1` means that there is only the initial value in `theta_list` where the initial value is the estimated policy value of MIPS when we use all available feature dimensions. 

After the initial value is added to `theta_list`, there is the `continue` statement of python in the loop, so the function outputs `theta_list[0]` only when `Eq. (22)` is true in the next iteration. 

Is this correct?

File: obp/ope/estimators_embed.py [hunk 3]
```diff
@@ -0,0 +1,666 @@
...
+
+    def _estimate_w_x_e(
+        self,
+        context: np.ndarray,
+        action: np.ndarray,
+        action_embed: np.ndarray,
+        pi_b: np.ndarray,
+        pi_e: np.ndarray,
+    ) -> np.ndarray:
+        """Estimate the marginal importance weights."""
```
[the remark this answers is not in the document]
Comment (usaito) [reply obp/ope/estimators_embed.py:191]: Thanks for pointing this out. First, this is not related to B-OPE. I implemented a procedure to estimate the marginal importance weight `w(x,e)` described in Section 3.3 of https://arxiv.org/pdf/2202.06317.pdf

In order to estimate `w(x,e)`, we first estimate `\pi(a|x,e)`, which is `p_a_e` in the code. Then, we take the expectation of `w(x,a)` with respect to `\hat{\pi}(a|x,e)`. (Eq.(2) of the paper).

[the remark this answers is not in the document]
Comment (usaito) [reply obp/ope/estimators_embed.py:191]: > 1. Linear Logistic model cannot capture interactions between contexts and actions

Thank you! That might be true. I modified the implementation a bit, and now users can choose any classifier such as RandomForest as `p_a_e_model`. (Actually, I also renamed `p_a_e_model` to `p_a_e_estimator`).


> 2. B-OPE should use Binary classification, but multi-class classification seems to be..

This is not B-OPE, but rather estimates `\pi(a|x,e)` via multi-class classification.

> 3. I cannot understand the last summation ( w_x_e = (w_x_a * p_a_e).sum(1) )

This takes the expectation of w(x,a) with respect to \hat{\pi}(a|x,e). (Eq.(2) of the paper).

**Comment by usaito** [comment]
@fullflu Thanks for the detailed feedback!

File: obp/ope/estimators_embed.py (hunk 2 above)
Comment (fullflu) [reply obp/ope/estimators_embed.py:366]: OK, I misunderstand the implementation!!

That might be correct and has no problem!

File: obp/ope/estimators_embed.py (hunk 1 above)
Comment (fullflu) [reply obp/ope/estimators_embed.py]: > theta_j[-1] is the last element of theta_j (or equivalently the last element of theta_list), this is not necessarily the element having the smallest CDF.

That is OK!
In my understanding, the feature extraction is implemented as the following steps:

1. The last element of `idx_list` indicate a feature that has  the smallest CDF among the `current_feature` (`idx_list = np.argsort(cnf_list_)[::-1]`)
2. The last element of `idx_list` is excluded if not returned (`idx_without_d = np.where(current_feat != excluded_dim, True, False)`)
3. That is why a feature that has the smallest CDF seems to be excluded.

Are the steps correct?
And I cannot understand why the last element of `idx_list` is excluded in the step3.

File: obp/ope/estimators_embed.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (fullflu) [reply obp/ope/estimators_embed.py:203]: OK, I understand the implementation and modification!!

@usaito 

[imo]

It would be better to add some (performance) tests when `p_e_a` is None and `embedding_selection_method` is not None in order to check the validity of the implementations.

File: obp/ope/estimators_embed.py (hunk 2 above)
Comment (usaito) [reply obp/ope/estimators_embed.py:366]: @fullflu thanks!

File: obp/ope/estimators_embed.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (usaito) [reply obp/ope/estimators_embed.py:203]: @fullflu Thanks for checking! Do you mean I should modify https://github.com/st-tech/zr-obp/blob/feature/marginal-ips/tests/ope/test_offline_estimation_performance.py ?

[the remark this answers is not in the document]
Comment (fullflu) [reply obp/ope/estimators_embed.py:203]: > Do you mean I should modify https://github.com/st-tech/zr-obp/blob/feature/marginal-ips/tests/ope/test_offline_estimation_performance.py ?

Yes!

Or, it would be sufficient to add estimators where `embedding_selection_method` is not None in: https://github.com/st-tech/zr-obp/blob/feature/marginal-ips/tests/ope/test_ipw_estimators_embed.py

File: obp/ope/estimators_embed.py (hunk 1 above)
Comment (usaito) [reply obp/ope/estimators_embed.py]: @fullflu Thanks for clarifying your point!

> The last element of idx_list indicate a feature that has the smallest CDF among the current_feature

Let me first clarify that the last element of `idx_list` is the feature that leads to the smallest CDF **when it is excluded from OPE**. And if all features pass the criterion `(np.abs(theta_j - theta_i) <= cnf_i + C * cnf_j).all()`, we know that we can exclude any feature from `current_feature` and proceed to the next round. Then, I think we should exclude the feature that leads to the smallest CDF, as it has already satisfied the criterion. (if this is not true, what you think is the appropriate feature to exclude when proceed to the next iteration?)

File: obp/ope/estimators_embed.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (usaito) [reply obp/ope/estimators_embed.py:203]: @fullflu Sure! will do from now

File: obp/ope/estimators_embed.py (hunk 1 above)
Comment (fullflu) [reply obp/ope/estimators_embed.py]: Thank you for your thorough explanation!
I read the algorithm again, and I find that the process has no problem!

(I think that excluding the feature that leads to  the **smallest CDF** or the **largest CDF** might be a **greedy** algorithm, and we cannot decide which one is better.)

Comment (usaito) [reply obp/ope/estimators_embed.py]: @fullflu Thanks for checking again! 

> the smallest CDF or the largest CDF might be a greedy algorithm, and we cannot decide which one is better.)

I agree, but I thought the feature that leads to the smallest CDF should be removed because passing `(np.abs(theta_j - theta_i) <= cnf_i + C * cnf_j).all()` becomes the most difficult with the smallest CDF, and if the feature with the smallest CDF pass the criterion, then we can be aggressive and simply remove that feature.
